src/framework/utils.hpp

```diff
@@ -697,6 +707,105 @@ matrix<T> tensor_product(const matrix<T> &A, const matrix<T> &B) {
   return temp;
 }
 
+template <class T>
+matrix<T> concatenate (const matrix<T> &A, const matrix<T> &B, uint_t axis) {
+  // Works out a concatenate of two matrices A, B
+  // To add B as new rows axis == 0 (vertical)
+  // To add A as new cols axis == 1 (horizontal)
+  if (axis != 0 && axis!= 1) {
+    std::cerr << "MU::concatenate: axis must be 0 or 1" << std::endl;
+	exit(1);
+  }
+  size_t rows1 = A.GetRows(), rows2 = B.GetRows(), cols1 = A.GetColumns(), cols2 = B.GetColumns();
+  matrix<T> temp = A;
+  if(axis == 0) {
+	if(cols1 != cols2) {
+	  std::cerr << "MU::concatenate: axis = 0 but cols_A =! cols_B" << std::endl;
+	  exit(1);
+	}
+  temp.resize(rows1 + rows2, cols1);
+  for (size_t i = 0; i < rows2; i++)
+	for (size_t j = 0; j < cols1; j++)
+      temp(rows1 + i,j) = B(i,j);
+  }
+  else if(axis == 1) {
+    if(rows1 != rows2) {
+	  std::cerr << "MU::concatenate: axis = 1 but rows_A =! rows_B" << std::endl;
+	  exit(1);
+	}
+	temp.resize(rows1, cols1 + cols2);
+	for (size_t i = 0; i < rows1; i++)
+	  for (size_t j = 0; j < cols2; j++)
+		temp(i,cols1 + j) = B(i,j);
+  }
+  return temp;
+}
+
+template <class T>
+void split (const matrix<T> &A, matrix<T> &B, matrix<T> &C, uint_t axis) {
+  // Works out a split of matrix A to two matrices B, C, with the same dimensions
+  // To split by rows, axis == 0 (Only when A rows divide by 2)
+  // To split by cols, axis == 1 (Only when A cols divide by 2)
+  if (axis != 0 && axis != 1) {
+    std::cerr << "MU::split: axis must be 0 or 1" << std::endl;
+    exit(1);
+  }
+  size_t rows = A.GetRows(), cols = A.GetColumns();
+  matrix<T> temp = A;
+  if(axis == 0) {
+	if (rows % 2 != 0) {
+	  std::cerr << "MU::split: can't matrix A split by rows"  << std::endl;
+	  exit(1);
+	}
+  B.resize(rows/2 , cols);
+  C.resize(rows/2 , cols);
+  for (size_t i = 0; i < rows/2; i++)
+	for (size_t j = 0; j < cols; j++) {
+	  B(i,j) = A(i,j);
+	  C(i,j) = A(i+rows/2,j);
+   	}
+  }
+  else if(axis == 1) {
+    if (cols % 2 != 0) {
+	  std::cerr << "MU::split: can't split matrix A by cols"  << std::endl;
+	  exit(1);
+	}
+	B.resize(rows, cols/2);
+	C.resize(rows, cols/2);
+	for (size_t i = 0; i < rows; i++)
```

Comment:
To avoid misunderstanding, I'd put the braces in this `for` loop:
```
for (size_t i = 0; i < rows; i++) {
```

Reply:
I agree.